src/cursor.rs

```diff
@@ -66,15 +67,124 @@ impl RawCursor {
 
     fn received_bytes(&self) -> u64 {
         match &self.0 {
-            RawCursorInner::Waiting(_) => 0,
-            RawCursorInner::Loading(state) => state.net_size,
+            RawCursorState::Waiting(_) => 0,
+            RawCursorState::Loading(state) => state.net_size,
         }
     }
 
     fn decoded_bytes(&self) -> u64 {
         match &self.0 {
-            RawCursorInner::Waiting(_) => 0,
-            RawCursorInner::Loading(state) => state.data_size,
+            RawCursorState::Waiting(_) => 0,
+            RawCursorState::Loading(state) => state.data_size,
+        }
+    }
+}
+
+// === BytesCursor ===
+
+/// Unlike [`RowCursor`] which emits rows deserialized as structures from RowBinary,
+/// this cursor emits raw bytes without deserialization.
+/// It can be iterated over using [`futures::StreamExt::next`].
+/// Additionally, if the requested format emits each row on a newline
+/// (e.g. `JSONEachRow`, `CSV`, `TSV`, etc.), the cursor can be converted into a stream of lines
+/// using [`futures::AsyncBufReadExt::lines`]`, which will allow for more convenient processing.
+#[cfg(feature = "futures03")]
+pub struct BytesCursor {
+    raw: RawCursor,
+    buffer: Vec<u8>,
+    filled: usize,
+    consumed: usize,
+}
+
+#[cfg(feature = "futures03")]
+impl BytesCursor {
+    pub(crate) fn new(response: Response) -> Self {
+        Self {
+            raw: RawCursor::new(response),
+            buffer: Vec::with_capacity(INITIAL_BYTES_CURSOR_BUFFER_SIZE),
+            filled: 0,
+            consumed: 0,
+        }
+    }
+}
+
+#[cfg(feature = "futures03")]
+impl futures::stream::Stream for BytesCursor {
+    type Item = Result<Bytes>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        use futures::FutureExt;
+        match self.raw.next().boxed().poll_unpin(cx) {
```

Comment:
This code relies on the fact that `fn next()` is cancel-safe. It seems to be cancel-safe for now, but more explicit way to implement such things is to provide `RawCursor::poll_next()`.

Also, boxing is extra allocation here (but probably doesn't affect performance)

Reply:
I tried to rewrite, and this unexpectedly affects the `select` benchmark (3.3GiB/s -> 2.2 GiB/s). Sadly, cachegrind and perf show nothing useful.